File: examples/basic/code_under_test.cpp
/**
 * Copyright 2018-2020 Pejman Ghorbanzade. All rights reserved.
 */

#include "code_under_test.hpp"
#include <thread>
#include <unordered_map>
#include <weasel/weasel.hpp>

/**
 *
 */
Wizard parse_profile(const std::string& path)
{
    static std::unordered_map<std::string, Wizard> wizards = {
        { "rweasley",
            { "rweasley",
                L"Ronald Weasley",
                Date { 1980, 3, 1 },
                6.3f,
                152.0,
                { "unicorn tail hair" } } },
        { "hpotter",
            { "hpotter",
                L"Harry James Potter",
                Date { 1980, 6, 31 },
                6.0f,
                145.0,
                { "phoenix feather" } } },
        { "hgranger",
            { "hgranger",
                L"Hermione Jean Granger",
                Date { 1979, 9, 19 },
                5.5f,
                120.0,
                { "dragon heartstring" } } }
    };

    return wizards.at(path);
}

/**
 *
 */
void custom_function_1(const Wizard& wizard)
{
    WEASEL_SCOPED_TIMER;
    weasel::add_result("is_tall", (6.0f < wizard.height));
    std::this_thread::sleep_for(std::chrono::milliseconds(10 + rand() % 50));
}

/**
 *
 */
void custom_function_2(const Wizard& wizard)
{
    for (auto i = 0ul; i < wizard.wands.size(); ++i) {
        weasel::scoped_timer timer("func2_wand_" + std::to_string(i));
        weasel::add_hit_count("number of wands");
        std::this_thread::sleep_for(std::chrono::milliseconds(10 + rand() % 50));
    }
}

/**
 *
 */
void custom_function_3(const Wizard& wizard)
{
    for (const auto& wand : wizard.wands) {
        weasel::add_array_element("wands", wand);
    }
    std::this_thread::sleep_for(std::chrono::milliseconds(10 + rand() % 50));
}

File: examples/js/packages/minimal/src/code_under_test.ts
/**
 * Copyright 2018-2020 Pejman Ghorbanzade. All rights reserved.
 */

export interface Wizard {
  username: string;
  fullname: string;
  dob: Record<'year' | 'month' | 'day', number>;
  height: number;
  weight: number;
  wands: string[];
}

const wizards: Wizard[] = [
  {
    username: 'rweasley',
    fullname: 'Ronald Weasley',
    dob: { year: 1980, month: 3, day: 1 },
    height: 6.3,
    weight: 152.0,
    wands: ['unicorn tail hair']
  },
  {
    username: 'hpotter',
    fullname: 'Harry Potter',
    dob: { year: 1980, month: 6, day: 31 },
    height: 6.0,
    weight: 145.0,
    wands: ['phoenix feather']
  },
  {
    username: 'hgranger',
    fullname: 'Hermione Jean Granger',
    dob: { year: 1979, month: 9, day: 19 },
    height: 5.5,
    weight: 120.0,
    wands: ['dragon heartstring']
  }
];

export function parse_profile(username: string): Wizard | undefined {
  return wizards.find((item) => item.username === username);
}
